File: src/MagnusAddon/qt_utils/i_task_progress_runner.py
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Callable


class ITaskRunner:
    def process_with_progress[TInput, TOutput](self, items: list[TInput], process_item: Callable[[TInput], TOutput], message: str, run_gc: bool = False, minimum_items_to_gc: int = 0) -> list[TOutput]: raise NotImplementedError()  # pyright: ignore
    def set_label_text(self, text: str) -> None: raise NotImplementedError()  # pyright: ignore
    def close(self) -> None: raise NotImplementedError()
    def run_on_background_thread_with_spinning_progress_dialog[TResult](self, message: str, action: Callable[[], TResult]) -> TResult: raise NotImplementedError()  # pyright: ignore
    def run_gc(self) -> None: pass
    def is_hidden(self) -> bool: return True

File: src/src_dotnet/JAStudio.Core/TaskRunners/ITaskProgressRunner.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Compze.Utilities.SystemCE.ActionFuncHarmonization;

namespace JAStudio.Core.TaskRunners;

public interface ITaskProgressRunner : IDisposable
{
   void RunBatch<TInput>(List<TInput> items, Action<TInput> processItem, string message) => RunBatch(items, processItem.AsFunc(), message, ThreadCount.One);
   void RunBatch<TInput>(List<TInput> items, Action<TInput> processItem, string message, ThreadCount threads) => RunBatch(items, processItem.AsFunc(), message, threads);
   public List<TOutput> RunBatch<TInput, TOutput>(List<TInput> items, Func<TInput, TOutput> processItem, string message) => RunBatch(items, processItem, message, ThreadCount.One);
   List<TOutput> RunBatch<TInput, TOutput>(List<TInput> items, Func<TInput, TOutput> processItem, string message, ThreadCount threads);

   Task<List<TOutput>> RunBatchAsync<TInput, TOutput>(List<TInput> items, Func<TInput, TOutput> processItem, string message) => RunBatchAsync(items, processItem, message, ThreadCount.One);
   Task<List<TOutput>> RunBatchAsync<TInput, TOutput>(List<TInput> items, Func<TInput, TOutput> processItem, string message, ThreadCount threadCount);

   Task RunBatchAsync<TInput>(List<TInput> items, Action<TInput> processItem, string message) => RunBatchAsync(items, processItem.AsFunc(), message, ThreadCount.One);
   Task RunBatchAsync<TInput>(List<TInput> items, Action<TInput> processItem, string message, ThreadCount threadCount) => RunBatchAsync(items, processItem.AsFunc(), message, threadCount);

   void RunIndeterminate(string message, Action action) => RunIndeterminate(message, action.AsFunc());
   TResult RunIndeterminate<TResult>(string message, Func<TResult> action);
   Task<TResult> RunIndeterminateAsync<TResult>(string message, Func<TResult> action);
   Task RunIndeterminateAsync(string message, Action action) => RunIndeterminateAsync(message, action.AsFunc());

   void SetLabelText(string text);
   bool IsHidden();
}
   